derive_builder_core/src/setter.rs

```diff
@@ -166,42 +166,56 @@ impl<'a> ToTokens for Setter<'a> {
 
 // adapted from https://stackoverflow.com/a/55277337/469066
 // Note that since syn is a parser, it works with tokens.
-// We cannot know for sure that this is an Option. The user could, for example, type
-// - `std::option::Option`
-// - `type MaybeString = std::option::Option<String>`
+// We cannot know for sure that this is an Option.
+// The user could, for example, `type MaybeString = std::option::Option<String>`
 // We cannot handle those arbitrary names.
 fn extract_type_from_option(ty: &syn::Type) -> Option<&syn::Type> {
-    use syn::{GenericArgument, Path, PathArguments};
-    fn path_is_option(path: &Path) -> bool {
-        path.leading_colon.is_none()
-            && path.segments.len() == 1
-            && path.segments.iter().next().unwrap().ident == "Option"
+    use syn::punctuated::Pair;
+    use syn::token::Colon2;
+    use syn::{GenericArgument, Path, PathArguments, PathSegment};
+
+    fn extract_type_path(ty: &syn::Type) -> Option<&Path> {
+        match *ty {
+            syn::Type::Path(ref typepath) if typepath.qself.is_none() => Some(&typepath.path),
+            _ => None,
+        }
     }
 
-    match *ty {
-        syn::Type::Path(ref typepath)
-            if typepath.qself.is_none() && path_is_option(&typepath.path) =>
-        {
-            // Get the first segment of the path (there is only one, in fact: "Option"):
-            typepath
-                .path
-                .segments
-                .first()
-                .and_then(|pair_path_segment| {
-                    let type_params = &pair_path_segment.value().arguments;
-                    // It should have only on angle-bracketed param ("<String>"):
-                    match *type_params {
-                        PathArguments::AngleBracketed(ref params) => params.args.first(),
-                        _ => None,
-                    }
-                })
-                .and_then(|generic_arg| match *generic_arg.into_value() {
-                    GenericArgument::Type(ref ty) => Some(ty),
-                    _ => None,
-                })
-        }
-        _ => None,
+    // TODO store (with lazy static) precomputed parsing of Option when support of rust 1.18 will be removed (incompatible with lazy_static)
```

Comment:
I'm fine merging as-is, but do you know if there's an older version of `lazy_static` we could use that is compatible with `rustc 1.18`? Unfortunately I don't have a good way of knowing the versions of Rust where this crate is being used, so I don't have a timetable for dropping this support.

Reply:
I tried without success the latest version of lazy_static and the version 1.0.0. using lazy_static was a better optimization  for a previous implementation I made.
TBH, the support 1.18 generate some trouble, I develop.